.github/workflows/gradle.yml

```diff
@@ -26,7 +25,30 @@ jobs:
       uses: actions/checkout@v2
     - name: Build with Gradle
       run: ./gradlew build
+    - name: Upload Unit Test Results
+      if: always()
+      uses: actions/upload-artifact@v3
+      with:
+        name: data-prepper-test-results-java-${{ matrix.java }}
+        path: '**/test-results/**/*.xml'
     - name: Upload Coverage Report
       uses: codecov/codecov-action@v1
     - name: Generate Javadocs
       run: ./gradlew javadoc
+
+  publish-test-results:
+    name: "Publish Unit Tests Results"
+    needs: build
+    runs-on: ubuntu-latest
+    if: always()
+
+    steps:
+      - name: Download Artifacts
+        uses: actions/download-artifact@v3
+        with:
+          path: test-results
+
+      - name: Publish Unit Test Results
+        uses: EnricoMi/publish-unit-test-result-action@v1
+        with:
+          files: "test-results/**/*.xml"
```

Comment:
Trying to understand the flow here. Why do we need to download artifacts ? How is step in line 28 different from line 51 ?

Reply:
@dinujoh , Great question.

The Publish Unit Test Result action publishes a clean summary of results from the JUnit results XML files. For a matrix strategy (which this build uses) the documentation notes that builds should upload the archives and then run a single job afterward to publish them. This is because it will aggregate all the XML files (from each matrix build).

[EnricoMi/publish-unit-test-result-action](https://github.com/EnricoMi/publish-unit-test-result-action#use-with-matrix-strategy)

Additionally, by uploading this XML, this becomes an attached artifact to the build. Developers can download the XML file and view the results locally on their machines. This can be useful in some situations.

Reply:
You can see the results of this in a build on my branch.

If you look at the bottom of [this build](https://github.com/dlvenable/data-prepper/actions/runs/2228323859), you will see an Artifacts section. You can download the XML from the [data-prepper-test-results-java-14](https://github.com/dlvenable/data-prepper/suites/6269385383/artifacts/223884563) link. This is what the upload action does.

The publish test results action creates what you can [see here](https://github.com/dlvenable/data-prepper/runs/6180716810?check_suite_focus=true).